### Goal

We have collected a number of tweets and dumped them in a database. If not, check out the "gather-twitter-info-example" notebook. We want to get an idea what people in the network are talking about, but we don't want to go through the tweets manually. Let's use the magic of machine learning instead: We learn a very simple text classifier on example texts and classify the tweets into these categories. 

### Train classifier

Let's train the text classifier on ~18000 newsgroup posts on 20 different topics (http://scikit-learn.org/stable/datasets/twenty_newsgroups.html). This data can be easily loaded from the scikit-learn library.  

#### Load data

In [12]:
%load_ext autoreload
%autoreload 2
from sklearn.datasets import fetch_20newsgroups
 
twenty_train = fetch_20newsgroups(subset='train', shuffle=True, random_state=32)
twenty_test = fetch_20newsgroups(subset='test', shuffle=True, random_state=32)

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


#### Train classifier

In the best case scenario this text classifier should be able to classify tweets from the pre-filtered twitter stream directly. So speed does matter. Therefore let's use a simple linear classifier: A multinomial naive bayesian classifier (http://scikit-learn.org/stable/modules/naive_bayes.html). To be able to classify text we first want to create a sparse vector representation of the documents (http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html) and tf-idf transform (https://en.wikipedia.org/wiki/Tf%E2%80%93idf) to take relative term frequency (tf) and inverse document frequency (idf) into account (http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfTransformer.html). 

We combine the CountVectorizer, TfidfTransformer and Classifier into a scikit-learn pipeline and identify the optimal set of meta parameters using 5-fold cross validation combined with a grid search. The parameters we are optimizing for are i) n-gram length for feature extraction (1 word or 1 and 2 words) and ii) the additive smoothing term for the naive bayes classifier. 

In [13]:
from src.message_classifier import MessageClassifier
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.naive_bayes import MultinomialNB

tc = MessageClassifier()
pipeline=Pipeline([('vect', CountVectorizer()), ('tfidf', TfidfTransformer()), ('nb', MultinomialNB())])
param_grid={'vect__ngram_range': [(1, 1),(1, 2)], 'nb__alpha': [10**-5,10**-4,10**-3,10**-2,10**-1]}
tc.train(train_X=twenty_train.data, train_y=twenty_train.target, labels=twenty_train.target_names, pipeline=pipeline, param_grid=param_grid)

tc.test(valid_X=twenty_test.data, valid_y=twenty_test.target) # Test performace on hold-out set

Fitting 5 folds for each of 10 candidates, totalling 50 fits


[Parallel(n_jobs=-1)]: Done  42 tasks      | elapsed:  6.8min
[Parallel(n_jobs=-1)]: Done  50 out of  50 | elapsed:  8.4min finished


Best score 0.917093866007 of estimator Pipeline(steps=[('vect', CountVectorizer(analyzer=u'word', binary=False, decode_error=u'strict',
        dtype=<type 'numpy.int64'>, encoding=u'utf-8', input=u'content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 2), preprocessor=None, stop_words=None,
        st...alse,
         use_idf=True)), ('nb', MultinomialNB(alpha=0.001, class_prior=None, fit_prior=True))])
                          precision    recall  f1-score   support

             alt.atheism       0.82      0.82      0.82       319
           comp.graphics       0.69      0.74      0.71       389
 comp.os.ms-windows.misc       0.73      0.62      0.67       394
comp.sys.ibm.pc.hardware       0.64      0.76      0.69       392
   comp.sys.mac.hardware       0.82      0.82      0.82       385
          comp.windows.x       0.83      0.79      0.81       395
            misc.forsale       0.86      0.83      0.84       390
               rec

Cross validation reports an average f1-score (weighted average of precission and recall) of 0.XX. This is classification performance is slighty too optimistic: The prediction on the hold out set reports an f1-score of 0.84.

#### Save the classifier
To be able to use the classifier in different applications. Let's save it. 